### DISCUSSION TITLE
aspカーネルのサービスコールのIDチェックに関して

### DISCUSSION THREAD

**Discussion by Ryohsuke-Takeuchi** [toppers/users-forum#153] [Q&A] [answered]
aspカーネルのソースを参照していて、以下のマクロの判定が間違っているのに気が付きました。
余りに基本的な間違いなので、本当なのか、何度も確認しましたが、
asp3カーネルでは、修正されているようなので、連絡します。
CHECK_TSKID_SELF
CHECK_TPRI_INI
CHECK_TPRI_SELF

CHECK_TSKID_SELFはタスクIDとTSK_SELFを判定するマクロですが、
私の間違いでしょうか？
確認をおねがいします
現状：
#define CHECK_TSKID_SELF(tskid) do {						\
	if (!(VALID_TSKID(tskid) || (tskid) == TSK_SELF)) {		\
		ercd = E_ID;										\
		goto error_exit;									\
	}														\
} while (false)
正しいと思われるマクロ：
#define CHECK_TSKID_SELF(tskid) do {						\
	if (!(VALID_TSKID(tskid) && (tskid) != TSK_SELF)) {		\
		ercd = E_ID;										\
		goto error_exit;									\
	}														\
} while (false)

**Comment by hiroakitakada** [comment, marked as answer]
`!`のかかっている範囲の関係で読みにくいマクロになっていますが，元の通りで正しいと考えます。
おそらく，`!`のかかっている範囲を読み違えられているものと思います。

まず，`VALID_TSKID(tskid)`は，tskidが正しい範囲（1〜最大値）に真になります。
次に，`VALID_TSKID(tskid) || (tskid) == TSK_SELF`で，tskidが正しい範囲か0になっている場合に真になります。
最後に，それに`!`がかかっていますので，（tskidが正しい範囲か0）でない場合に真になります。

> 正しいと思われるマクロ：
> #define CHECK_TSKID_SELF(tskid) do {
> if (!(VALID_TSKID(tskid) && (tskid) != TSK_SELF)) {
> ercd = E_ID;
> goto error_exit;
> }
> } while (false)

`!`がかかっている範囲が`VALID_TSKID(tskid)`のみであれば，`&& (tskid) != TSK_SELF`が来るのが正しいですが，`!`が`VALID_TSKID(tskid) || (tskid) == TSK_SELF`全体にかかっているので，元のプログラムで正しいです。

**Reply by Ryohsuke-Takeuchi** [reply]
すみません
今、ソースを見て取り下げの返事を書こうと思って、開いたら回答が記載されていて申し訳ありません。
先生のおっしゃる通りです。
エラー条件で、判定すると、以下のマクロになりますね。
休みの日にすみません。

#define CHECK_TSKID_SELF(tskid) do {						\
	if ( !VALID_TSKID(tskid) && (tskid) != TSK_SELF ) {		\
		ercd = E_ID;										\
		goto error_exit;									\
	}														\
} while (false)
